README.md
<p align="center"><img src="https://www.holbertonschool.com/holberton-logo.png"></p>

-----
# SIMPLE SHELL

Simple shell is a fundamental command line user interface where you wich provides all basic built-in commands.
This project is our own simple shell with basic commands that was created based on SH. Simple shell handle environment
variables in order to improve functionality of commands such as cd. There are two built-in functions to handle those
environment variables (setenv, unsetenv).

-----
## INSTALLATION

1. Git clone this repository

```bash
 shell:~/root/simple_shell$ git clone https://github.com/andergcp/simple_shell.git
```

2. Compile all files to get executable file

```bash
 shell:~/root/simple_shell$ gcc *.c -o custom_name_shell
```

3. Execute

```bash
 shell:~/root/simple_shell$ ./custom_name_shell
```

-----
## USAGE

### INTERACTIVE MODE

Interactive mode allows you to enter as many commands as you want.
Also allows you to exit in two ways, press ```CTRL + D``` or
exit built-in command. ```$ exit ```

To run simple shell just execute the command below

```bash
 shell:~/root/simple_shell$ ./custom_name_shell
```

It will display this screen that means you are inside
```bash
 $
```
```bash
shell:~/root/simple_shell$ ./custom_name_shell
 $ echo test
 test
 $
```

### NON INTERACTIVE MODE

Non interactive mode allows you to run a simple command.

```bash
 shell:~/root/simple_shell$ echo "command to run" | ./custom_name_shell
```

```bash
 shell:~/root/simple_shell$ echo "ls -l" | ./custom_name_shell
 -rwxrw-r-- 1 vagrant vagrant    30 Feb 20 00:00 custom_name_shell
```

-----
## BUILT-IN COMMANDS

More information about each command use help. ```$ help built-in_command```


```bash
shell:~/root/simple_shell$ ./custom_name_shell
 $ help exit
 Exit the shell.

    Exits the shell with a status of N.  If N is omitted, the exit status
    is that of the last command executed.
 $
```

| Command | Description  |
| ------- | --- |
| exit | Exit simple shell|
| help | Display help of built-in command |
| env | Display environment variables |
| cd | Change directory |
| setenv | Set environment variable value or create a new one |
| unsetenv | Delete environment variable |

-----
## EXIT STATUS

There are different posible exit status, this deppende on the success of the last command.

0: Success exit.<br>
2: Error execute command. <br>
126: Permission denied.<br>
127: not found.

-----
## EXAMPLES

### INTERACTIVE MODE

```bash
 shell:~/root/simple_shell$ ./custom_name_shell
 $ ls -l
 -rwxrw-r-- 1 vagrant vagrant    30 Feb 20 00:00 custom_name_shell
 $
```
```bash
 shell:~/root/simple_shell$ ./custom_name_shell
 $ echo 123
 123
 $
```

```bash
 shell:~/root/simple_shell$ ./custom_name_shell
 $ cd /tmp
 $ pwd
 /tmp
 $
```

### NON INTERACTIVE MODE

```bash
 shell:~/root/simple_shell$ echo "/bin/ls -l" | ./custom_name_shell
 -rwxrw-r-- 1 vagrant vagrant    30 Feb 20 00:00 custom_name_shell
 shell:~/root/simple_shell$
 $
```
```bash
 shell:~/root/simple_shell$ echo "pwd" | ./custom_name_shell
 /root/simple_shell
 shell:~/root/simple_shell$
 $
```

```bash
 shell:~/root/simple_shell$ echo "exit 7" | ./custom_name_shell
 shell:~/root/simple_shell$ echo $?
 shell:~/root/simple_shell$ 7
```

-----
## FILES

Help_files: Directory containing help files of built-in commands.

| File | Description  |
| ------- | --- |
| shell.h | Declaration of all functions ande include of libraries|
| main.c | Main file with function to start: <br> - main() <br> - initialize()  |
| basic_fun_1.c | Function to handle commands and errors: <br> handle_path() <br> _getptr() <br> _getoken()  <br> _execute() <br> manage_command()  |
| basic_fun_2.c | Functions to support basic functions: <br> dic_command() <br> error_msg() <br> strtok_line() <br> strtok_path()  |
| aux_fun_1.c | Auxiliar functions: <br> _strlen() <br> _strcat() <br> _strcmp() <br> _strdup() <br> strcpy() |
| aux_fun_2.c | Auxiliar functions: <br> clear_path() <br> atoi() <br> itoa() <br> c_buf() |
| built_func_1.c | Built-in functions: <br> comm_cd() <br> comm_he() <br> comm_ex() <br> comm_en() |
| built_func_2.c | Built-in functions: <br> comm_unset() <br> comm_set() |
| env_fun | Functions to handle environment variables: <br> set_env_v() <br> clear_env() <br> set_env() <br> get_env() |

## Authors
- John Garcia  [@jgarcia15121982](https://www.github.com/jgarcia15121982)
- Anderson Castiblanco [@andergcp](https://www.github.com/andergcp)
- Alejandro Castiblanco [@AlejoCasti](https://www.github.com/AlejoCasti)


help_files/history.txt
history: history [-c] [-d offset] [n] or history -anrw [filename] or history -ps arg [arg...]
    Display or manipulate the history list.

    Display the history list with line numbers, prefixing each modified
    entry with a `*'.  An argument of N lists only the last N entries.

    Options:
      -c        clear the history list by deleting all of the entries
      -d offset delete the history entry at offset OFFSET.

      -a        append history lines from this session to the history file
      -n        read all history lines not already read from the history file
      -r        read the history file and append the contents to the history
        list
      -w        write the current history to the history file
        and append them to the history list

      -p        perform history expansion on each ARG and display the result
        without storing it in the history list
      -s        append the ARGs to the history list as a single entry

    If FILENAME is given, it is used as the history file.  Otherwise,
    if $HISTFILE has a value, that is used, else ~/.bash_history.

    If the $HISTTIMEFORMAT variable is set and not null, its value is used
    as a format string for strftime(3) to print the time stamp associated
    with each displayed history entry.  No time stamps are printed otherwise.

    Exit Status:
    Returns success unless an invalid option is given or an error occurs.


basic_func.c
#include "shell.h"
/**
 * dic_command - create a new diccionary
 * Return: pointer that point to new diccionary
 */
inside *dic_command()
{
	inside *array_comm = malloc(sizeof(inside) * 8);
	
	if (array_comm)
	{
		array_comm[0].command = "cd";
		array_comm[0].command_function = comm_cd;
		array_comm[0].help = "cd.txt";
		array_comm[1].command = "help";
		array_comm[1].command_function = comm_he; 
		array_comm[1].help = "help.txt";
		array_comm[2].command = "exit";
		array_comm[2].command_function = comm_ex;
		array_comm[2].help = "exit.txt";
		array_comm[3].command = "env";
		array_comm[3].command_function = comm_en;
		array_comm[3].help = "env.txt";
		array_comm[4].command = "history";
		array_comm[4].command_function = comm_his;
		array_comm[4].help = "history.txt";
		array_comm[5].command = "unsetenv";
		array_comm[5].command_function = comm_unset;
    array_comm[5].help = "unsetenv.txt";
		array_comm[6].command = "setenv";
		array_comm[6].command_function = comm_set;
    array_comm[6].help = "setenv.txt";
		array_comm[7].command = NULL;
		array_comm[7].command_function = NULL;
    array_comm[7].help = NULL;

		return (array_comm);
	}
	return (NULL);
}
/**
 * _getptr - get new line
 * Return: pointer that point to new line
 */
void _getptr(variables *m_v)
{
	size_t i;
	int x;
	char *ptr = NULL;

	x = getline(&ptr, &i, stdin);
	if (x > 0)
	{
		m_v->ptr = ptr;
		return;
	}
	else
	{
		if (isatty(0))
			write(STDOUT_FILENO, "\n", 1);
		free(ptr);
		m_v->ptr = NULL;
	}
}
/**
 * _getoken - split line into tokens
 * @ptr: pointer of line
 * Return; array of tokens
 */
void _getoken(variables *m_v)
{
	char *ptraux = NULL;
	int i = 0;

	if (!(m_v->ptr[0]) == '\n')
		return;
	m_v->args = _strtok_line(m_v->ptr);
	if (!(m_v->args))
		return ;
}
/**
 * _execute - execute command
 * @args: arrays of tokens
 * @argv: arguments of main
 */
void _execute(variables *m_v, char *args)
{
	pid_t f_pid;
	struct stat aux_stat;
	int status;
	DIR *dir = NULL;

	if (args == NULL)
	{
		dir = opendir(m_v->args[0]);
		if (dir && (_strcmp(m_v->args[0], ".") != 0))
		{
			error_msg(m_v, "Permission denied"), m_v->status = 126;
			closedir(dir);
			return;
		}
		else if (stat(m_v->args[0], &aux_stat) == -1)
		{
			error_msg(m_v, "not found"), m_v->status = 127;
			return;
		}
		else if (access(m_v->args[0], X_OK) == -1)
		{
			error_msg(m_v, "Permission denied"), m_v->status = 126;
			return;
		}
		args = m_v->args[0];
	}
	f_pid = fork();
	if (f_pid == -1)
		error_msg(m_v, "fork failed");
	if (f_pid == 0)
	{
		execve(args, m_v->args, NULL);
		free(m_v->args);
		_exit(2);
	}
	waitpid(f_pid, &status, WUNTRACED);
	m_v->status = WEXITSTATUS(status);
}
/**
 * manage_command - handles the command line search
 * @m_v: general struct
 * Return: 0 in right return, and -1 when failed
 */
int manage_command(variables *m_v)
{
	int i;
	char *hp_arg = NULL;
	struct stat aux_stat;
	
	if (m_v->diccio)
	{
		for (i = 0; m_v->diccio[i].command; i++)
			if (_strcmp(m_v->diccio[i].command, m_v->args[0]) == 0)
			{				
				i = m_v->diccio[i].command_function(m_v);
				return (i);
			}
	}
	else
		return (-1);
	hp_arg = handle_path(m_v);
	_execute(m_v, hp_arg);
	free(hp_arg);
	return (0);
}
